pydantic/_internal/_discriminated_union.py

```diff
@@ -430,6 +430,18 @@ def _infer_discriminator_values_for_inner_schema(
             # This will happen if the field has a default value; we ignore it while extracting the discriminator values
             return self._infer_discriminator_values_for_inner_schema(schema['schema'], source)
 
+        elif schema['type'] == 'function-after':
+            # After validators don't affect the discriminator values
+            return self._infer_discriminator_values_for_inner_schema(schema['schema'], source)
+
+        elif schema['type'] in {'function-before', 'function-wrap', 'function-plain'}:
```

Comment:
Not sure how you got around typing here but might be worth using `_core_utils.is_function_with_inner_schema(`

Reply:
Note that it's not actually equivalent to that function call. We could do something like that, but this works because I'm not actually using any fields from inside `schema` to generate this error, so I don't need to do any type narrowing of `schema`.

The reason I did this was because each of those types of validator could affect what goes in, whereas `function-after` can't, so it's safe to rely on the input discriminator to select the union case to use for type-checking.